src/analysis/processing/qgsalgorithmlinedensity.cpp

```diff
@@ -0,0 +1,216 @@
+
+    //default weight of lines is set to 1 if no field provided
+    double analysisWeight = 1;
+    if ( !mWeightField.isEmpty() )
+    {
+      analysisWeight = f.attribute( mWeightField ).toDouble();
+    }
+    mFeatureWeights.insert( f.id(), analysisWeight );
+  }
+
+
+  const QString outputFile = parameterAsOutputLayer( parameters, QStringLiteral( "OUTPUT" ), context );
+  QFileInfo fi( outputFile );
+  const QString outputFormat = QgsRasterFileWriter::driverForExtension( fi.suffix() );
+
+  int rows = std::max( std::ceil( mExtent.height() / mPixelSize ), 1.0 );
+  int cols = std::max( std::ceil( mExtent.width() / mPixelSize ), 1.0 );
+
+  //build new raster extent based on number of columns and cellsize
+  //this prevents output cellsize being calculated too small
+  QgsRectangle rasterExtent = QgsRectangle( mExtent.xMinimum(), mExtent.yMaximum() - ( rows * mPixelSize ), mExtent.xMinimum() + ( cols * mPixelSize ), mExtent.yMaximum() );
+
+  QgsRasterFileWriter writer = QgsRasterFileWriter( outputFile );
+  writer.setOutputProviderKey( QStringLiteral( "gdal" ) );
+  writer.setOutputFormat( outputFormat );
+  std::unique_ptr<QgsRasterDataProvider > provider( writer.createOneBandRaster( Qgis::Float32, cols, rows, rasterExtent, mCrs ) );
+  if ( !provider )
+    throw QgsProcessingException( QObject::tr( "Could not create raster output: %1" ).arg( outputFile ) );
+  if ( !provider->isValid() )
+    throw QgsProcessingException( QObject::tr( "Could not create raster output %1: %2" ).arg( outputFile, provider->error().message( QgsErrorMessage::Text ) ) );
+
+  provider->setNoDataValue( 1, -9999 );
+
+  qgssize totalCellcnt = static_cast<qgssize>( rows ) * cols;
+  int cellcnt = 0;
+
+  for ( int row = 0; row < rows; row++ )
+  {
+    std::unique_ptr< QgsRasterBlock > rasterDataLine = qgis::make_unique< QgsRasterBlock >( Qgis::Float32, cols, 1 );
+    for ( int col = 0; col < cols; col++ )
+    {
+      if ( feedback->isCanceled() )
+      {
+        break;
+      }
+
+      if ( col > 0 )
+        mSearchGeometry.translate( mPixelSize, 0 );
+
+      const QList<QgsFeatureId> fids = mIndex.intersects( mSearchGeometry.boundingBox() );
+
+      if ( !fids.isEmpty() )
+      {
+        std::unique_ptr< QgsGeometryEngine > engine( QgsGeometry::createGeometryEngine( mSearchGeometry.constGet() ) );
+        engine->prepareGeometry();
+
+        double absDensity = 0;
+        for ( QgsFeatureId id : fids )
+        {
+          QgsGeometry lineGeom = mIndex.geometry( id );
+
+          if ( engine->intersects( lineGeom.constGet() ) )
+          {
+            double analysisLineLength =  mDa.measureLength( QgsGeometry( engine->intersection( mIndex.geometry( id ).constGet() ) ) );
+            double weight = mFeatureWeights.value( id );
+            absDensity += ( analysisLineLength *  weight );
+          }
+        }
+
+        //use ellipsoidal area
+        double analysisSearchGeometryArea = mDa.measureArea( mSearchGeometry );
```

Comment:
Another optimisation -- if absDensity is zero here, we could skip the area measurement calculation since the pixel value will always be 0.